### ISSUE TITLE
XmlSerializer: Cannot serialize member because it is an interface

### ISSUE THREAD

**Issue by vanillajonathan** [dotnet/runtime#81214]
### Description

`XmlSerializer` cannot deserialize XML into a class where the member is a collection interface.

### Reproduction Steps

```csharp
[XmlRoot("Document", Namespace = "urn:example:test")]
public class Document
{
    public ICollection<Pet> Pets { get; init; } = null!; // ICollection does not work, but List works
}
```

### Expected behavior

I expected it to be able to deserialize XML into a class with collection members that are interfaces, such as `IEnumerable<T>`, `ICollection<T>`, `IReadOnlyCollection<T>` and `IList<T>`.

### Actual behavior

> NotSupportedException: Cannot serialize member Example.Document.Pets of type System.Collections.Generic.ICollection`1[[Example.Pets, Example, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]] because it is an interface.

### Known Workarounds

I have to deserialize into `List<T>` or `Pet[]`.

### Configuration

.NET 7
Windows 11 - 21H2 / 22000.1516
x64

**Comment by pvlst** [comment]
Can you please make this a priority? If you want to deserialize XML code that contains a collection, the only choice is to either use
- array property => triggers warning [CA1819](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1819)
- List property => triggers warning [CA1002](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1002)

So basically, you cannot write a clean code now. Or am I wrong?

**Comment by StephenMolloy** [comment]
As I'm sure you know, `XmlSerializer` is unable to deserialize properties typed as interfaces (such as `ICollection<T>`, `IEnumerable<T>`, or `IReadOnlyCollection<T>`) because interfaces do not provide sufficient information for object instantiation. During deserialization, `XmlSerializer` needs to construct an object to hold the incoming data. However, interfaces in .NET are abstract contracts—they don’t specify how to create an actual instance, nor do they guarantee the presence of a default constructor. As a result, `XmlSerializer` cannot know what concrete implementation to use, and so it throws an error when it encounters a property of interface type.

This limitation is rooted in how `XmlSerializer` operates: it relies on being able to create instances of your types using default constructors. When you use a concrete type like `List<T>` or `Pet[]`, `XmlSerializer` knows precisely how to instantiate that collection and fill it with elements. With an interface type, there’s no such guarantee, which is why serialization and deserialization of interface properties are not supported natively.

As a workaround, you can use a concrete “backing” property specifically for serialization, while keeping your public API typed to the interface. For example, you can mark your interface property with `[XmlIgnore]` and introduce a corresponding concrete property (e.g., `List<Pet>`) with `[XmlElement]` or `[XmlArray]` attributes for `XmlSerializer` to use. You can then map between the interface and the concrete type in your property's getter and setter. This approach maintains compatibility with analyzers and best practices by keeping your public API abstracted, while still allowing `XmlSerializer` to work with a serializable concrete type.

Here’s a quick example:

```csharp
[XmlIgnore]
public ICollection<Pet> Pets { get; set; } = new List<Pet>();

[XmlArray("Pets")]
[XmlArrayItem("Pet")]
public List<Pet> PetsBacking
{
    get => Pets.ToList();
    set => Pets = value;
}
```

This technique allows your model to adhere to best practices (using interfaces for public properties) while still being compatible with `XmlSerializer`'s requirements. It’s a bit of boilerplate, but it’s currently the most reliable way to bridge the gap between modern .NET design and the limitations of `XmlSerializer`.

I recognize this is not an ideal answer, but it is the general case answer. Given that the current exception happens before deserialization, I'm not sure how much flexibility we have to do anything differently or if any common cases (`IList<>` for example) might be able to find some more palatable workaround.

**Comment by vanillajonathan** [comment]
I think whenever it comes across a property with a interface either is `IEnumerable<T>` or inherits from `IEnumerable<T>` then it should instantiate it as a `List<T>`.

EF Core and System.Text.Json are able to work with classes where the properties are interfaces of collections.
